userspace/libsinsp/sinsp_resource_utilization.cpp

```diff
@@ -0,0 +1,156 @@
+				vsz /= 1024;		/* convert to MB with precision loss by design */
+			} else if(strstr(line, "VmRSS:") == line)
+			{
+				sscanf(line, "VmRSS: %u", &rss);		/* memory size returned in kb */
+				rss /= 1024;		/* convert to MB with precision loss by design */
+			}
+		}
+		fclose(f);
+	}
+
+    snprintf(filename, sizeof(filename), "/proc/%d/smaps_rollup", pid);
+	if(FILE* f = fopen(filename, "r"))
+	{
+		while(fgets(line, sizeof(line), f) != NULL)
+		{
+			if(strstr(line, "Pss:") == line)
+			{
+				sscanf(line, "Pss: %u", &pss);		/* memory size returned in kb */
+				pss /= 1024;		/* convert to MB with precision loss by design */
+				break;
+			}
+		}
+		fclose(f);
+	}
+}
+
+void libsinsp::resource_utilization::get_cpu_usage(double &cpu_usage_perc, const scap_agent_info* agent_info)
+{
+
+	struct tms time;
+	if (times (&time) != (clock_t) -1)
+	{
+
+		/* Number of clock ticks per second, often referred to as USER_HZ / jiffies. */
+		long hz = 100;
+#ifdef _SC_CLK_TCK
+		if ((hz = sysconf(_SC_CLK_TCK)) < 0)
+		{
+			ASSERT(false);
+		}
+#endif
+		struct sysinfo s_info;
+		if(sysinfo(&s_info) != 0)
+		{
+			ASSERT(false);
+		}
+
+		/* Current utime is amount of processor time in user mode of calling process.
+		 * Convert to seconds. */
+		double user_sec = (double)time.tms_utime / hz;
+
+		/* Current stime is amount of time the calling process has been scheduled in kernel mode.
+		 * Convert to seconds. */
+		double system_sec = (double)time.tms_stime / hz;
+
+		/* Current uptime of the host machine from sysinfo in seconds. */
+		long machine_uptime_sec = s_info.uptime;
+
+		/* CPU usage as percentage is computed by dividing the time the process uses the CPU by the
+		 * currently elapsed time of the calling process. Compare to `ps` linux util. */
+		double elapsed_sec = machine_uptime_sec - agent_info->start_time;
+		cpu_usage_perc = (double)100.0 * (user_sec + system_sec) / elapsed_sec;
+		cpu_usage_perc = std::round(cpu_usage_perc * 10.0) / 10.0; // round to 1 decimal
+	}
+}
+
+void libsinsp::resource_utilization::get_container_memory_usage(uint32_t &memory_used)
+{
+	/* In Kubernetes `container_memory_working_set_bytes` is the memory measure the OOM killer uses
+	 * and values from `/sys/fs/cgroup/memory/memory.usage_in_bytes` are close enough.
+	 * -> contrasted numbers from multiple sources in a real-life Kubernetes cluster.
+	 *
+	 * Please note that `kubectl top pod` numbers would reflect the sum of containers in a pod and
+	 * typically libs clients (e.g. Falco) pods contain sidekick containers that use memory as well.
+	 * This metric accounts only for the container with the security monitoring agent running.
+	*/
+	char filename[1024];
+	const char *env_filename = getenv(SINSP_AGENT_CGROUP_MEM_PATH_ENV_VAR);
+	if(env_filename != NULL)
+	{
+		snprintf(filename, sizeof(filename), "%s", env_filename);
+	}
+	else
+	{
+		snprintf(filename, sizeof(filename), "%s", "/sys/fs/cgroup/memory/memory.usage_in_bytes");
```

Comment:
The statement that this is close enough to the Kubernetes `container_memory_working_set_bytes` which the OOM killer uses (see Kubernetes [metrics, metrics-server](https://kubernetes.io/docs/tasks/configure-pod-container/assign-memory-resource/)) is peer-review by infrastructure folks from my organization. We compared the numbers from Kubernetes prometheus endpoints with memory.usage_in_bytes after execing into a running container ...